### ISSUE TITLE
Retained MQTT messages just partly recieved on restart/deploy

### ISSUE THREAD

**Issue by Alex9779** [node-red/node-red#1950]
- [x] Node-RED version: 0.19.4
- [x] node.js version: 9.2.1
- [x] npm version: 5.6.0
- [x] Platform/OS: Debian Stretch
- [x] Browser: Chrome/Firefox

I worked the whole day to this point and I hope I find some help here.
As the title states I have a problem with retained MQTT messages. Just a part of all retained messages are received by Node-RED when I restart or when I deploy after a change. "Just a part" is not exactly right, they also get multiplied. So some are missing, some are multiple times there and some are missing completely.
I tried this with multiple versions of node-red (0.17.5, 0.18.1, 0.18.7) and at first I thought this was depending on the version, later on I discovered it seems to be dependent on the number of MQTT nodes using the same MQTT server. Ok not the same server but the same server config node.
Because if I just create a new server entry for the same server and just use that in one MQTT node then it works as intended.
To show what I mean with some logs here is what I get when I subscribe the topic "homie/+/$name" on my mosquitto MQTT broker with the mosquitto_sub command:
```
Bed Uplight Left
Living Room Switch
Spielzimmer_Flügel
Couch Uplight Right
NH-3-Relay
Bed Uplight Right
Flur Relay
Couch Uplight Left
Die ist ein Test
NH 2 Relay
```
And this is what I get when I subscribe the same topic in Node-RED with a MQTT server config node that is only used by this one MQTT subscribe node:
```
0|node-red | 31 Oct 08:33:57 - [info] [debug:b10f5aaa.203a68] Bed Uplight Left
0|node-red | 31 Oct 08:33:57 - [info] [debug:b10f5aaa.203a68] Living Room Switch
0|node-red | 31 Oct 08:33:57 - [info] [debug:b10f5aaa.203a68] Spielzimmer_Flügel
0|node-red | 31 Oct 08:33:57 - [info] [debug:b10f5aaa.203a68] Couch Uplight Right
0|node-red | 31 Oct 08:33:57 - [info] [debug:b10f5aaa.203a68] NH-3-Relay
0|node-red | 31 Oct 08:33:57 - [info] [debug:b10f5aaa.203a68] Bed Uplight Right
0|node-red | 31 Oct 08:33:57 - [info] [debug:b10f5aaa.203a68] Flur Relay
0|node-red | 31 Oct 08:33:57 - [info] [debug:b10f5aaa.203a68] Couch Uplight Left
0|node-red | 31 Oct 08:33:57 - [info] [debug:b10f5aaa.203a68] Die ist ein Test
0|node-red | 31 Oct 08:33:57 - [info] [debug:b10f5aaa.203a68] NH 2 Relay
```
Those two are identical!
But now when I use a MQTT server config node which is use by 66 nodes (in and out):
```
0|node-red | 31 Oct 08:35:59 - [info] [debug:21e954ff.a833ac] Living Room Switch
0|node-red | 31 Oct 08:35:59 - [info] [debug:b10f5aaa.203a68] Bed Uplight Left
0|node-red | 31 Oct 08:35:59 - [info] [debug:b10f5aaa.203a68] Living Room Switch
0|node-red | 31 Oct 08:35:59 - [info] [debug:b10f5aaa.203a68] Spielzimmer_Flügel
0|node-red | 31 Oct 08:35:59 - [info] [debug:b10f5aaa.203a68] Couch Uplight Right
0|node-red | 31 Oct 08:35:59 - [info] [debug:b10f5aaa.203a68] NH-3-Relay
0|node-red | 31 Oct 08:35:59 - [info] [debug:b10f5aaa.203a68] Bed Uplight Right
0|node-red | 31 Oct 08:35:59 - [info] [debug:b10f5aaa.203a68] Flur Relay
0|node-red | 31 Oct 08:35:59 - [info] [debug:b10f5aaa.203a68] Couch Uplight Left
0|node-red | 31 Oct 08:35:59 - [info] [debug:b10f5aaa.203a68] Die ist ein Test
0|node-red | 31 Oct 08:35:59 - [info] [debug:b10f5aaa.203a68] NH 2 Relay
0|node-red | 31 Oct 08:35:59 - [info] [debug:21e954ff.a833ac] Couch Uplight Left
0|node-red | 31 Oct 08:35:59 - [info] [debug:85a8c3f3.f37748] Flur Relay
0|node-red | 31 Oct 08:35:59 - [info] [debug:21e954ff.a833ac] Couch Uplight Right
0|node-red | 31 Oct 08:35:59 - [info] [debug:85a8c3f3.f37748] Spielzimmer_Flügel
0|node-red | 31 Oct 08:35:59 - [info] [debug:85a8c3f3.f37748] Bed Uplight Left
0|node-red | 31 Oct 08:35:59 - [info] [debug:85a8c3f3.f37748] Living Room Switch
0|node-red | 31 Oct 08:35:59 - [info] [debug:85a8c3f3.f37748] Spielzimmer_Flügel
0|node-red | 31 Oct 08:35:59 - [info] [debug:85a8c3f3.f37748] Couch Uplight Right
0|node-red | 31 Oct 08:35:59 - [info] [debug:85a8c3f3.f37748] NH-3-Relay
0|node-red | 31 Oct 08:35:59 - [info] [debug:85a8c3f3.f37748] Bed Uplight Right
0|node-red | 31 Oct 08:35:59 - [info] [debug:85a8c3f3.f37748] Flur Relay
0|node-red | 31 Oct 08:35:59 - [info] [debug:21e954ff.a833ac] Bed Uplight Left
0|node-red | 31 Oct 08:35:59 - [info] [debug:85a8c3f3.f37748] Couch Uplight Left
0|node-red | 31 Oct 08:35:59 - [info] [debug:85a8c3f3.f37748] Die ist ein Test
0|node-red | 31 Oct 08:35:59 - [info] [debug:85a8c3f3.f37748] NH 2 Relay
```
And this with a config node used by 101 other nodes:
```
0|node-red | 31 Oct 08:36:48 - [info] [debug:b10f5aaa.203a68] Bed Uplight Left
0|node-red | 31 Oct 08:36:48 - [info] [debug:b10f5aaa.203a68] Living Room Switch
0|node-red | 31 Oct 08:36:48 - [info] [debug:b10f5aaa.203a68] Spielzimmer_Flügel
0|node-red | 31 Oct 08:36:48 - [info] [debug:b10f5aaa.203a68] Couch Uplight Right
0|node-red | 31 Oct 08:36:48 - [info] [debug:b10f5aaa.203a68] NH-3-Relay
0|node-red | 31 Oct 08:36:48 - [info] [debug:b10f5aaa.203a68] Bed Uplight Right
0|node-red | 31 Oct 08:36:48 - [info] [debug:b10f5aaa.203a68] Flur Relay
0|node-red | 31 Oct 08:36:48 - [info] [debug:b10f5aaa.203a68] Couch Uplight Left
0|node-red | 31 Oct 08:36:48 - [info] [debug:b10f5aaa.203a68] Die ist ein Test
0|node-red | 31 Oct 08:36:48 - [info] [debug:b10f5aaa.203a68] NH 2 Relay
0|node-red | 31 Oct 08:36:48 - [info] [debug:85a8c3f3.f37748] Flur Relay
0|node-red | 31 Oct 08:36:48 - [info] [debug:21e954ff.a833ac] Living Room Switch
0|node-red | 31 Oct 08:36:48 - [info] [debug:85a8c3f3.f37748] Spielzimmer_Flügel
0|node-red | 31 Oct 08:36:48 - [info] [debug:21e954ff.a833ac] Couch Uplight Left
0|node-red | 31 Oct 08:36:48 - [info] [debug:85a8c3f3.f37748] Bed Uplight Left
0|node-red | 31 Oct 08:36:48 - [info] [debug:85a8c3f3.f37748] Living Room Switch
0|node-red | 31 Oct 08:36:48 - [info] [debug:85a8c3f3.f37748] Spielzimmer_Flügel
0|node-red | 31 Oct 08:36:48 - [info] [debug:85a8c3f3.f37748] Couch Uplight Right
0|node-red | 31 Oct 08:36:48 - [info] [debug:85a8c3f3.f37748] NH-3-Relay
0|node-red | 31 Oct 08:36:48 - [info] [debug:85a8c3f3.f37748] Bed Uplight Right
0|node-red | 31 Oct 08:36:48 - [info] [debug:85a8c3f3.f37748] Flur Relay
0|node-red | 31 Oct 08:36:48 - [info] [debug:85a8c3f3.f37748] Couch Uplight Left
0|node-red | 31 Oct 08:36:48 - [info] [debug:85a8c3f3.f37748] Die ist ein Test
0|node-red | 31 Oct 08:36:48 - [info] [debug:85a8c3f3.f37748] NH 2 Relay
0|node-red | 31 Oct 08:36:48 - [info] [debug:21e954ff.a833ac] Couch Uplight Right
0|node-red | 31 Oct 08:36:48 - [info] [debug:21e954ff.a833ac] Bed Uplight Left
```
The order is always the same. What you can also see is that even if I disabled the other debug nodes they still seem to give output.
If I just use the same debug node then the output is as follows:
using config node used by 66 nodes:
```
0|node-red | 31 Oct 09:23:21 - [info] [debug:21e954ff.a833ac] Flur Relay
0|node-red | 31 Oct 09:23:21 - [info] [debug:21e954ff.a833ac] Spielzimmer_Flügel
0|node-red | 31 Oct 09:23:21 - [info] [debug:21e954ff.a833ac] Bed Uplight Left
0|node-red | 31 Oct 09:23:21 - [info] [debug:21e954ff.a833ac] Living Room Switch
0|node-red | 31 Oct 09:23:21 - [info] [debug:21e954ff.a833ac] Spielzimmer_Flügel
0|node-red | 31 Oct 09:23:21 - [info] [debug:21e954ff.a833ac] Couch Uplight Right
0|node-red | 31 Oct 09:23:21 - [info] [debug:21e954ff.a833ac] NH-3-Relay
0|node-red | 31 Oct 09:23:21 - [info] [debug:21e954ff.a833ac] Bed Uplight Right
0|node-red | 31 Oct 09:23:21 - [info] [debug:21e954ff.a833ac] Flur Relay
0|node-red | 31 Oct 09:23:21 - [info] [debug:21e954ff.a833ac] Couch Uplight Left
0|node-red | 31 Oct 09:23:21 - [info] [debug:21e954ff.a833ac] Die ist ein Test
0|node-red | 31 Oct 09:23:21 - [info] [debug:21e954ff.a833ac] NH 2 Relay
```
And using the config node used by 101:
```
0|node-red | 31 Oct 09:25:56 - [info] [debug:21e954ff.a833ac] Living Room Switch
0|node-red | 31 Oct 09:25:56 - [info] [debug:21e954ff.a833ac] Couch Uplight Left
0|node-red | 31 Oct 09:25:56 - [info] [debug:21e954ff.a833ac] Couch Uplight Right
0|node-red | 31 Oct 09:25:56 - [info] [debug:21e954ff.a833ac] Bed Uplight Left

```
This is just one of multiple retained topics my devices set. For example for the IP address, that is the topic I really need and because of which I had to dig into this problem, topics are multiplied or completely missing.
For a temp workaround I created a second server config node so most of my system works as intended with that split of 101 to 66 after a deploy or restart but still something strange is going on here.
I searched the web and here but I didn't find any advice or limit regarding the maximum of MQTT nodes using the same server config node but obviously I have a problem with that.

So any ideas?

BTW, all MQTT server config nodes have the exact same settings...

**Comment by knolleary** [comment]
The key question is: for the config nodes that have multiple mqtt nodes using them, what topics are those nodes subscribed to?

If you have two mqtt nodes subscribed to the same topic, that will cause two separate subscriptions being sent to the broker, both of which will trigger the retained messages to be sent - you'll get one copy of them per mqtt subscriber. When they arrive at the mqtt config node, it finds all mqtt nodes whose subscription matches and passes the message one. That means not only are multiple copies sent to the config node, the config node will send each copy to every matching node.

Does that fit with what you see?

**Comment by Alex9779** [comment]
Ok as I have a lot of flows, some equal becaue the devices are equal, this is a tricky question to answer.
As far as I can tell my subscribing nodes regarding the topics from my devices are mostly unique because I get that event from the device and handle it, in a handful of cases there are two or three nodes subscribing the same topic.
What I have multiple times subscribe are the topics coming from homebridge-mqtt, there are multiple in and out nodes subscribing the same topic in one flow.
But for the example I created above the $name topic is subscribed anywhere else.
And this is the one and only subscription I use wildcards.
I also tested it with explicit topics, that is also what happens to my nodes subscribing the ip address topic, there is one node in most of the flows subscribing the explicit ip address topic of one device, some of those won't get the ip address upon deployment, others do.

So to sum up, no that is not what I see I think.

**Comment by Alex9779** [comment]
No problem, I will try what I can to clarify... So lets start with some screenshots of how I produced the logs posted.
Here are my configuration nodes for the MQTT servers:
![image](https://user-images.githubusercontent.com/6394581/47783926-5e823780-dd04-11e8-82c9-e8798cb04f62.png)
This is how I produced the first output which is identical to what the direct MQTT sub told me:
![image](https://user-images.githubusercontent.com/6394581/47783977-8d001280-dd04-11e8-8ad8-9a1ca6a49647.png)
Here the second setup already givign me a mix of the now disabled debug node and the active one:
![image](https://user-images.githubusercontent.com/6394581/47784014-a608c380-dd04-11e8-9319-608b6994a99f.png)
And the last one:
![image](https://user-images.githubusercontent.com/6394581/47784030-b3be4900-dd04-11e8-9dd1-cb4a15c6e9d3.png)
To get the last two logs where I don't see the output from disabled debug nodes I deleted two debug nodes and just rewired every subscription node to teh one debug node like this:
![image](https://user-images.githubusercontent.com/6394581/47784088-e23c2400-dd04-11e8-8738-b7095f4064c8.png)

**Comment by knolleary** [comment]
I assume the (1)/(66)/(101) labels in the MQTT nodes represent which config node they are using?

**Comment by Alex9779** [comment]
Yup

**Comment by Alex9779** [comment]
Maybe just an idea, I have mosquitto with persitence activated, so when I quit it I have the db file with the retained topics. I could provide this file and also a full backup of my flows.
As I said, it seems not dependent on the installation because I tried this with my imported lfow on multiple instances with different versions of Node-RED (docker images) and always the same behaviour.
With the mosquitto db file it should be possible to run a duplicate server in a test environment and with the flows it should be possible to repoduce...
But I don't really wanna post this public accessible here in the thread...
Just an idea...

**Comment by Alex9779** [comment]
Ok I think I am getting closer... I had some other wildcard subscription nodes around in the flows.
And now I better understand your first post...
So one subscription is for "homie/+/$name" and another is for "homie/nh-2-relay/#" while the first one is wired to a debug node, second one is not bound to anything, I see the output for "homie/nh-2-relay/$name" twice in the output...
Is that by design?
Things screw up them I have a subscription for "homie/#" also, then the output shortens and is missing a lot of items...

**Comment by Alex9779** [comment]
I debugged and digged into the code though it is hard for my a bit unexperienced with node.js, the style of the 10-mqtt.js is.... unsteady at least in the sections I am looking at. Still trying to understand how the code is executed but let me just share some thoughts.
I think I understand it now a little better an can answer my own question of the post before thought I can't answer really if that behaviour is by design meaning intentional but it is at least unavaiodable at the moment. As far as I can see is the MQTT lib designed to manage a client MQTT connection but like for one user. You can subscribe a topic and that's it, then you get the messages if there arrive some for that topic. The subscription is just on the basis of the topic. But Node-RED is a little different, you can have multiple nodes, each subscribing the same or eqal topic paths leading to the same topic when using wildcards. The MQTT lib doesn't care and just pushes the messages and you just have the topic you can compare, nothing else, so you HAVE TO send that message to all node subscribing that topic. That explains the duplication I described in the post above.
I still don't know why a # wildcard screws that up.
IMHO this is a really strange behaviour for someone getting Node-RED, seeing all that nice nodes, placing them, no warning or description of that behaviour because as it showed in my first post even subscriptions on nodes that are wired to deactivated debug nodes still recieve that message. And here must be a problem within Node-RED because I get this messages on an active debug node which is wired to an MQTT node subscribing the same or an equal topic.
The real problem IMHO is that you are not able to bind the subscription to the node that subscribes becauue all you have is the topic. The underlying MQTT lib does not allow you to tag or give an addtional property when subscribing which is then returned in the callback function when a message arrives so you can handle that and post that message just to the node that subscribed.
At least I didn't find a possibility though the main README file states that you can pass a "properties" object with a "subscriptionIdentifier" and additional "userProperties". But the is no explanation what they are for or how they are used and no indication that they come back to you in the callback function. Also strange that this is jsut documented in the README file but not in the Wiki.

So to sum up I think we have some sort of a problem in Node-RED as described in the post before where a "#" subscription screws the others up (I cleared all my flows of other wildcard subscriptions before, and now there are definately no other wildcard subscriptions or subscriptions to the same topic around in my flows, still have to verify that on an empty Node-RED installation).
And the other thing is that the design how subscriptions are handled without being bound to the specific node that subscribed is at least questionable...

**Comment by knolleary** [comment]
Some of what you describe is what I was trying to explain in my original reply.

As a protocol, MQTT provides no way to identify which subscription caused any given message to arrive. That is why the underlying lib doesn't provide that behaviour and why we can't do anything about it in the NR nodes.

If you have overlapping subscriptions (`homie/+/$name` and `homie/nh-2-relay/#`) then the act of subscribing to each of those will trigger any retained messages to be sent. So that means you will get two copies. And each of those copies will get passed to any node that has a matching topic filter - so each node will get both copies.


> So one subscription is for "homie/+/$name" and another is for "homie/nh-2-relay/#" while the first one is wired to a debug node, second one is not bound to anything, I see the output for "homie/nh-2-relay/$name" twice in the output...

Yes. The node has subscribed to ``homie/+/$name` but the message was published to `homie/nh-2-relay/$name` so that is what its `topic` property is set to, so that is what gets displayed in Debug.

Aside from the duplicate messages due to overlapping subscriptions, I've not been able to recreate any issue having a node subscribed to # mixed in with others.

Here's what I've tested with:

<img width="757" alt="local_node-red" src="https://user-images.githubusercontent.com/51083/47879744-4d771a80-de19-11e8-8ea9-dbbb4397a44f.png">

Importable flow in the collapsed section here:

<details>
 <pre>[{"id":"4249d0d4.e8c64","type":"inject","z":"ac62deb3.6e136","name":"","topic":"","payload":"","payloadType":"date","repeat":"","crontab":"","once":false,"onceDelay":0.1,"x":120,"y":140,"wires":[["56ff4a0e.a5a594"]]},{"id":"56ff4a0e.a5a594","type":"function","z":"ac62deb3.6e136","name":"","func":"var rooms = \"abcde\".split(\"\");\nrooms.forEach(function(r) {\n    node.send({\n        retain: true,\n        topic: \"test/\"+r+\"/$name\",\n        payload: \"this is \"+r\n    })\n});\nreturn;","outputs":1,"noerr":0,"x":270,"y":140,"wires":[["6ee7afeb.d3718"]]},{"id":"6ee7afeb.d3718","type":"mqtt out","z":"ac62deb3.6e136","name":"","topic":"","qos":"","retain":"","broker":"10ae742b.8314ec","x":430,"y":140,"wires":[]},{"id":"ef468ab0.813ef8","type":"mqtt in","z":"ac62deb3.6e136","name":"","topic":"test/+/$name","qos":"2","datatype":"auto","broker":"10ae742b.8314ec","x":130,"y":280,"wires":[["c1fdcae4.c9c4f8"]]},{"id":"d794751b.4e4388","type":"mqtt in","z":"ac62deb3.6e136","name":"","topic":"test/#","qos":"2","datatype":"auto","broker":"10ae742b.8314ec","x":110,"y":340,"wires":[["c1fdcae4.c9c4f8"]]},{"id":"77a7c2ef.1a485c","type":"mqtt in","z":"ac62deb3.6e136","name":"","topic":"test/a/#","qos":"2","datatype":"auto","broker":"10ae742b.8314ec","x":110,"y":400,"wires":[["c1fdcae4.c9c4f8"]]},{"id":"46d2a256.dc1fec","type":"mqtt in","z":"ac62deb3.6e136","name":"","topic":"#","qos":"2","datatype":"auto","broker":"10ae742b.8314ec","x":110,"y":460,"wires":[["c1fdcae4.c9c4f8"]]},{"id":"de3e05ed.704028","type":"inject","z":"ac62deb3.6e136","name":"","topic":"","payload":"","payloadType":"date","repeat":"","crontab":"","once":false,"onceDelay":0.1,"x":120,"y":220,"wires":[["436761b5.3ec56"]]},{"id":"436761b5.3ec56","type":"change","z":"ac62deb3.6e136","name":"","rules":[{"t":"set","p":"counts","pt":"flow","to":"{}","tot":"json"}],"action":"","property":"","from":"","to":"","reg":false,"x":320,"y":220,"wires":[[]]},{"id":"c1fdcae4.c9c4f8","type":"function","z":"ac62deb3.6e136","name":"","func":"var counts = flow.get(\"counts\") || {};\ncounts[msg.payload] = (counts[msg.payload] || 0)+1\nflow.set(\"counts\",counts);\nreturn msg;","outputs":1,"noerr":0,"x":330,"y":340,"wires":[["86796bdc.9415a8"]]},{"id":"86796bdc.9415a8","type":"trigger","z":"ac62deb3.6e136","op1":"","op2":"counts","op1type":"nul","op2type":"flow","duration":"1","extend":true,"units":"s","reset":"","bytopic":"all","name":"","x":480,"y":340,"wires":[["50374ebb.0abb7"]]},{"id":"50374ebb.0abb7","type":"debug","z":"ac62deb3.6e136","name":"","active":true,"tosidebar":true,"console":false,"tostatus":false,"complete":"false","x":650,"y":340,"wires":[]},{"id":"ecad4233.6f102","type":"comment","z":"ac62deb3.6e136","name":"Setup retained - do this once!","info":"","x":160,"y":100,"wires":[]},{"id":"190923f2.2189dc","type":"comment","z":"ac62deb3.6e136","name":"Clear flow.counts","info":"","x":120,"y":180,"wires":[]},{"id":"10ae742b.8314ec","type":"mqtt-broker","z":"","name":"","broker":"localhost","port":"1883","clientid":"","usetls":false,"compatmode":false,"keepalive":"60","cleansession":true,"birthTopic":"","birthQos":"0","birthPayload":"","closeTopic":"","closeQos":"0","closePayload":"","willTopic":"","willQos":"0","willPayload":""}]
</pre>
</details>

The top flow puts some retained messages into the broker.

The second flow clears out the `flow.counts` object to an empty object.

The remaining nodes are subscribed to a variety of patterns. As messages arrive, the `flow.counts` object is updated to count how many of each payload is received. The final result is passed to debug.

There are three nodes that match all retained topics, so we expect 9 copies of them to be emitted (3 subscribers, so 3 copies of the retained messages passed to each of the 3 matching nodes - 3x3=9).
One node only matches `test/a/$name` - so 16 copies (4 copies to 4 nodes - 4x4=16).

And that's exactly what my testing gets:
```
{"this is a":16,"this is b":9,"this is c":9,"this is d":9,"this is e":9}
```

Is this a fair attempt at reproducing your scenario? Is there an aspect I'm missing?

**Comment by Alex9779** [comment]
*sigh* so I tried it again and now I am not able to reproduce the problem... I cleared all the flows, reimported from my production system, deleted all the wildcard nodes as before and created the nodes for subscribung $name... Nothing... Just working as you described, multiple replies because of not identifiable subscriptions...

Sorry for all that, I cleared all the the wildcards from my prod sys, removed all the addtional MQTT client setups, alls work again with one system node, and now it is working again without dropping messages I need...

**Comment by knolleary** [comment]
Thanks for the update. If you do hit anything like it again, please let us know!

**Comment by Alex9779** [comment]
Ya now problem... I assume because of some of the wildcard subs the parser got too many messages from the MQTT client and wasn't able to process all of them. Maybe that's why some were missing with all those subs up...
